### PR TITLE
To automatically restart from ADIOS output

### PR DESCRIPTION
**Pull request by BifengLei**
The `k80_profileRestart.tpl` file should be modified to automatically restart from ADIOS output.
Since only the last checkout is used for the automatic restart, the previous one, which maybe quite large, should be deleted in each step to save the disk space.

### CODE REVIEW

**Comment by ax3l** [comment]
Hi @BifengLei,

thank you for your first PR, fantastic! :sparkles:

We will start reviewing it and will add in-code annotations that we can discuss and improve before it gets merged to make it even more useful. Don't be afraid, we love to see your contribution and will assign you an experienced maintainer to lead the review process.

If you have not already done so, please cross-check our [contribution guidelines](https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/CONTRIBUTING.md). For further updates, please add them simply with additional commits to the branch which will update your pull request.

I will now add you to our "Developer" list to authorize you in our test suite as soon as you push the next time to this pull request's branch.

Again: welcome and thank you! :sparkles:

**Comment by PrometheusPi** [comment]
@BifengLei This replaces the `k80_profileRestart.tpl` designed for hdf5 by one for adios. Could you please submit it as a second `*.tpl`e.g.  `k80_profileRestartADIOS.tpl` so that both options are available.

**Comment by PrometheusPi** [comment]
The change looks good. Have already tested it on hypnos?

**Comment by ax3l** [comment]
Let us alternatively just detect if adios files are available.

Our checkpoint priorities are:
- if ADIOS & HDF5 are both available, ADIOS will take precedence for checkpoint writing and reading

**Comment by PrometheusPi** [comment]
I agree with @ax3l. One file would be better that automatically detects either hdf5 or adios restart files.

**Comment by PrometheusPi** [comment]
Please add your name to the author/contributor list in the header file.

File: src/picongpu/submit/hypnos-hzdr/k80_profileRestart.tpl [hunk 1]
```diff
@@ -77,10 +77,10 @@ echo "----- automated restart routine -----" | tee -a output
 
 #check whether last checkpoint is valid
 file=""
-for file in `ls -t ./checkpoints/checkpoint_*.h5`
```
Comment (ax3l) [inline src/picongpu/submit/hypnos-hzdr/k80_profileRestart.tpl]: suggestion to write a general file:

```bash
# ADIOS restart files take precedence over HDF5 files
fileEnding="h5"
hasADIOS=$(ls ./checkpoints/checkpoint_*.bp 2>/dev/null | wc -w)
if [ $hasADIOS -gt 0 ]
then
    fileEnding="bp"
fi

for file in `ls -t ./checkpoints/checkpoint_*.$fileEnding`
do
     echo -n "validate checkpoint $file: " | tee -a output
     $fileEnding"ls" $file &> /dev/null
```

File: src/picongpu/submit/hypnos-hzdr/k80_profileRestart.tpl [hunk 2]
```diff
@@ -102,7 +102,7 @@ echo  "restart period = " $restartPeriod | tee -a output
 
 if [ "" != "$file" ]
 then
-    cptimestep=`basename $file | sed 's/checkpoint_//g' | sed 's/.h5//g'`
+    cptimestep=`basename $file | sed 's/checkpoint_//g' | sed 's/.bp//g'`
```
Comment (ax3l) [inline src/picongpu/submit/hypnos-hzdr/k80_profileRestart.tpl]:
```bash
cptimestep=`basename $file | sed 's/checkpoint_//g' | sed 's/.'$fileEnding'//g'`
```

**Comment by ax3l** [comment]
> Since only the last checkout is used for the automatic restart, the previous one, which maybe quite large, should be deleted in each step to save the disk space.

That is true in most cases, but not necessarily all (ie if one wants to check a certain part in time of an existing simulation again with different plugins enabled, etc.). Nevertheless, feel free to remove all other files in a local copy of your .tpl file! :)

Did you know ADIOS supports [on-the-fly compression](https://github.com/ComputationalRadiationPhysics/picongpu/wiki/Plugin%3A-ADIOS#how-to-set-it-up)? ([Example](https://github.com/ComputationalRadiationPhysics/picongpu/blob/386700ad27549cea01e1c177d8cc18ded8ae58e5/docs/TBG_macros.cfg#L166))

**Comment by BifengLei** [comment]
@PrometheusPi  Yes. It has been tested on hypnos.
@ax3l Yes. I am curretly using this option for output..

File: src/picongpu/submit/hypnos-hzdr/k80_profileRestart.tpl [hunk 3]
```diff
@@ -77,10 +77,18 @@ echo "----- automated restart routine -----" | tee -a output
...
+if [ $hasADIOS -gt 0 ]
+then
+    fileEnding="bp"
+fi
+
+for file in `ls -t ./checkpoints/checkpoint_*.$fileEnding`
 do
-    echo -n "validate checkpoint $file: " | tee -a output
-    bpls $file &> /dev/null
+     echo -n "validate checkpoint $file: " | tee -a output
```
Comment (PrometheusPi) [inline src/picongpu/submit/hypnos-hzdr/k80_profileRestart.tpl]: REQUIRED: 
please remove the extra spaces

File: src/picongpu/submit/hypnos-hzdr/k80_profileRestart.tpl [hunk 4]
```diff
@@ -102,9 +110,9 @@ echo  "restart period = " $restartPeriod | tee -a output
 
 if [ "" != "$file" ]
 then
-    cptimestep=`basename $file | sed 's/checkpoint_//g' | sed 's/.bp//g'`
+    cptimestep=`basename $file | sed 's/checkpoint_//g' | sed 's/.'$fileEnding'//g'`
     echo "start time      = " $cptimestep | tee -a output
-
+
```
Comment (PrometheusPi) [inline src/picongpu/submit/hypnos-hzdr/k80_profileRestart.tpl]: REQUIRED:
please remove spaces

**Comment by PrometheusPi** [comment]
@BifengLei There still appears to be white space at the end of line. (See this travis-ci [output](https://travis-ci.org/ComputationalRadiationPhysics/picongpu/builds/203857901))

**Comment by PrometheusPi** [comment]
@BifengLei 
If you are using **emacs** - just run *Meta* x: `whitespace-cleanup`.
If you are using **vim** - just execute `:%s/\s\+$//e`.

**Comment by ax3l** [comment]
Our [Travis test](https://travis-ci.org/ComputationalRadiationPhysics/picongpu/builds/203857901) simply suggests to use sed: `sed -i 's/[[:blank:]]\+$//' ./src/picongpu/submit/hypnos-hzdr/k80_profileRestart.tpl`

File: src/picongpu/submit/hypnos-hzdr/k80_profileRestart.tpl [hunk 5]
```diff
@@ -102,11 +110,11 @@ echo  "restart period = " $restartPeriod | tee -a output
 
 if [ "" != "$file" ]
 then
-    cptimestep=`basename $file | sed 's/checkpoint_//g' | sed 's/.h5//g'`
+    cptimestep=`basename $file | sed 's/checkpoint_//g' | sed 's/.'$fileEnding'//g'`
     echo "start time      = " $cptimestep | tee -a output
 
     endTime="$(($cptimestep + $restartPeriod ))"
-    echo "end time        = " $endTime | tee -a output
+    echo "end time      = " $endTime | tee -a output
```
Comment (ax3l) [inline src/picongpu/submit/hypnos-hzdr/k80_profileRestart.tpl]: pls do not change this line :)

File: src/picongpu/submit/hypnos-hzdr/k80_profileRestart.tpl [hunk 6]
```diff
@@ -77,10 +77,18 @@ echo "----- automated restart routine -----" | tee -a output
...
+if [ $hasADIOS -gt 0 ]
+then
+    fileEnding="bp"
+fi
+
+for file in `ls -t ./checkpoints/checkpoint_*.$fileEnding`
 do
-    echo -n "validate checkpoint $file: " | tee -a output
-    h5ls $file &> /dev/null
+    echo -n "validate checkpoint $file:" | tee -a output
```
Comment (ax3l) [inline src/picongpu/submit/hypnos-hzdr/k80_profileRestart.tpl]: pls do not change this line :)

**Comment by PrometheusPi** [comment]
@BifengLei Could you add your name to the license header?

**Comment by ax3l** [comment]
@BifengLei thank you for the updates!
If you want, feel free to rebase this branch to `squash` or `fixup` all 6 commits into one.

You can do so by going in the branch, then:
```bash
git rebase -i HEAD~6
# editor opens
```

Change the lines from
```
pick 6592e6b To automatically restart from ADIOS output,  the `k80_profileRestart.tpl` file should be modified.
pick f00094b Automatic restart either from hdf5 or adios files.  ADIOS restart files take precedence over HDF5 files.
pick 56a9746 Remove some whitespace.
pick 3d11131 Remove the whitespace
pick a692e8e Modify the format for information print.
pick 0ae61ce Add Bifeng Lei to license header.
```

to
```
pick 6592e6b To automatically restart from ADIOS output,  the `k80_profileRestart.tpl` file should be modified.
s f00094b Automatic restart either from hdf5 or adios files.  ADIOS restart files take precedence over HDF5 files.
f 56a9746 Remove some whitespace.
f 3d11131 Remove the whitespace
f a692e8e Modify the format for information print.
f 0ae61ce Add Bifeng Lei to license header.
```

then save.
This merges all commits into one and opens a a new editor with the two upper commit messages so you can create a single, new cleaned up commit message.

Save that one again and then force push to update the PR:
```bash
git push -f
```

**Comment by ax3l** [comment]
wuhu, it worked! thanks!

**Comment by PrometheusPi** [comment]
@BifengLei Great :+1:
@ax3l Please restart travis-ci again - the checks passed but the GitHub status did not change.

**Pull request merged**